### DISCUSSION TITLE
x-for does not update browser if element nested in a reusable component

### DISCUSSION THREAD

**Discussion by peterswords** [alpinejs/alpine#3447] [7. Bugs (archived)]
When I use `x-for` over an array, nested elements update correctly when the array changes. However, if those nested elements are inside a reusable component defined using `Alpine.data`, they do *not* update.

In my real-world example these are complex nested components implementing expandable accordions and new data is fetched from a server. However, I've boiled it down to the bare essentials here so that the component definition is completely empty. A button click changes the content of the array from [x,y] to [a,b].

``` html
<!DOCTYPE html>
<html lang="en">
<head>
    <script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.js"></script>
</head>

<body x-data="{
    my_array: [{x:'x'},{x:'y'}],
    click() {
        this.my_array = [{x:'a'},{x:'b'}];
    }
}">

    <!-- Loop with plain div -->
    <template x-for="item in my_array">
        <div x-text="item.x"></div>
    </template>

    <!-- Loop with div nested inside component -->
    <template x-for="item in my_array">
        <div x-data="some_component" >
            <div x-text="item.x"></div>
        </div>
    </template>

    <button @click="click">Click me</button>

    <script>
        document.addEventListener('alpine:init', () => {
            Alpine.data('some_component', () => ({}));
        });

    </script>
</body>
</html>
```
JSFiddle is [here][1].
Initially the output looks like this, with two iterations over the array:

[![output before click][2]][2]

On clicking and replacing the array, the first iteration updates but the second does not:

[![output after click][3]][3]

----
**EDIT:**
If I parametrise the Alpine.data component with my data, everything works as expected:

``` HTML
...
    <template x-for="item in my_array">
        <div x-data="some_component(item)" >
            <div x-text="item.x"></div>
        </div>
    </template>
...
    <script>
        document.addEventListener('alpine:init', () => {
            Alpine.data('some_component', (i) => ({
                item: i,
            }));
        });

    </script>
```
Updated JSFiddle [here][4].

This gives me a workaround I can live with. However, it forces me to choose the number of parameters that will be passed to the component, and I may have to repackage certain data used within it. In my real-world example, a generic accordion component doesn't really care about the sort of layout or associated data that will live inside it. It would be preferable if the nested components could see "through" the `Alpine.data` component to data contained in outer scopes. And it's also weird that the nested components can see outer data when first rendered, but don't see changes to those data.

  [1]: https://jsfiddle.net/ps200306/zf3s8o45/3/
  [2]: https://i.stack.imgur.com/7ou41.png
  [3]: https://i.stack.imgur.com/LsBM4.png
  [4]: https://jsfiddle.net/ps200306/zf3s8o45/6/

**Comment by ekwoka** [comment]
Check with keys on the x-fors please. That is basically necessary for x-for and leaving it off will have strange behaviors

This should work fine even going through nested contexts, so check that first.

**Reply by peterswords** [reply]
I'm using keys in my real-world example, and also had them in the simple example above before I trimmed it down to barest essentials. It makes no difference to the problem.
Also, the workaround I came up with isn't working in my real-world example, I'm still trying to get to the bottom of why that is and produce another bare-bones version.

Here's a [JSFiddle](https://jsfiddle.net/ps200306/zf3s8o45/12/) using keys, which doesn't work.

**Reply by ekwoka** [reply]
that jsfiddle with this content:

```js
<!DOCTYPE html>
<html lang="en">
<head>
    <script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.js"></script>
</head>

<body x-data="{
    my_array: [{x:'x', k:1},{x:'y', k:2}],
    click() {
        this.my_array = [{x:'a', k:3},{x:'b', k:4}];
    }
}">

    <!-- Loop with plain div -->
    <template x-for="item in my_array" :key="item.k">
        <div x-text="item.x"></div>
    </template>

    <!-- Loop with div nested inside component -->
    <template x-for="item in my_array">
        <div x-data="{ item }" >
            <div x-text="item.x"></div>
        </div>
    </template>

    <button @click="click">Click me</button>

</body>
</html>
```

does work, so that's a start, or passing it through of course.

Very strange. Will look into it

directly updating properties on the objects works, and adding removing items works, but yeah, replacing the array or replacing objects in the array does not, like `item` is being memoized, and not updating the scope. May just need a simple effect added into `x-for` to handle that maybe.

So most operations should be pretty much fine, just not replacing the object with a new object that has the same key. That causes an issue.

Why do you need to do the operation like that though? Feels like you should be updating values in line.

**Reply by peterswords** [reply]
Actually, I got that last JSFiddle wrong -- I hadn't put the key on the second iteration, my mistake. I've now tried a number of different variations to see what works. I specify keys in all cases. Here's the basic code:

``` HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <script defer src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.js"></script>
</head>

<body x-data="{
    my_array: [{x:'x', k:1},{x:'y', k:2}],
    click() {
        this.my_array = [{x:'a', k:3},{x:'b', k:4}];
    }
}">

    <!-- Loop with plain div -->
    <template x-for="item in my_array" :key="item.k">
        <div x-text="item.x"></div>
    </template>

    <!-- Loop with div nested inside component -->
    <template x-for="item in my_array" :key="item.k">
        <div x-data="some_component(item)" >
            <div x-text="item.x"></div>
        </div>
    </template>

    <button @click="click">Click me</button>

    <script>
        document.addEventListener('alpine:init', () => {
            Alpine.data('some_component', () => ({
            }));
        });

    </script>
</body>
</html>
```

I've tried the following variations which differ only in the way I update the array:
1. Replace the entire array. The new objects have different keys to the old ones. This works. (`this.my_array = [{x:'a', k:3},{x:'b', k:4}];` [JSFiddle](https://jsfiddle.net/ps200306/zf3s8o45/24/)).
2. Replace the entire array. The new objects have the same keys as the old ones. Doesn't work. (`this.my_array = [{x:'a', k:1},{x:'b', k:2}];` [JSFiddle](https://jsfiddle.net/ps200306/zf3s8o45/26/)).
3. Replace objects in the array (but not the array itself).  The new objects have the same keys as the old ones. Doesn't work. (`this.my_array[0] = {x:'a', k:1}; this.my_array[1] = {x:'b', k:2};` [JSFiddle](https://jsfiddle.net/ps200306/zf3s8o45/28/)).
4. Replace objects in the array (but not the array itself).  The new objects have different keys to the old ones. This works. (`this.my_array[0] = {x:'a', k:3}; this.my_array[1] = {x:'b', k:4};` [JSFiddle](https://jsfiddle.net/ps200306/zf3s8o45/33/)).
5. Update fields of existing objects in the array (but not their keys). This works. (`this.my_array[0].x = 'a'; this.my_array[1].x = 'b';` [JSFiddle](https://jsfiddle.net/ps200306/zf3s8o45/31/)).

From this I draw the following tentative conclusions:
* If keys change Alpine will successfully update, regardless of whether you overwrite the array with an entirely new array object, or replace individual objects in the array.
* If keys don't change, Alpine will not update if you replace the array or objects in the array. However, if you update fields in place on existing objects in the array, it will update.

This is pretty awkward:
> Why do you need to do the operation like that though? Feels like you should be updating values in line.

I'm reading values from a remote database by doing a fetch on a server, and displaying the results in a table. If I refresh the UI and perform the fetch again, I have no idea what will change: some items may be added or deleted, or there may be updates to existing objects. All I have is a new array from fetch. Since many of my keys will not change (even though fields on those objects may have changed), it would appear from the experiments above that I have to look after my own mergesort on the existing array --  adding or removing objects where keys are new/deleted, but updating existing objects in place field by field. This would be a crazy amount of work and error prone since my schema may change and new fields get added. Maybe I'm missing something here?

However, the elephant in the room here is that the behaviour is _different_ between the code using a reusable component via `Alpine.data` and the code that doesn't. All of the above variations work when not using an `Alpine.data` component. Something very strange is going on.

**Reply by ekwoka** [reply]
I don't find it works if you inline it. It still doesn't work if the item itself isn't being passed to the child component.

So its in the process of "copying" the objects into the child scopes.

this may take some time to figure out

🤔

**Reply by peterswords** [reply]
Just to check I'm understanding -- "_inlining_" means not having an intervening `Alpine.data` component? I think that's how it's referred to in the documentation [here](https://alpinejs.dev/directives/data#re-usable-data), while the `Alpine.data` version is an _extracted_ component. If so, the inline version in all my five experiments above works ([1](https://jsfiddle.net/ps200306/zf3s8o45/24/), [2](https://jsfiddle.net/ps200306/zf3s8o45/26/), [3](https://jsfiddle.net/ps200306/zf3s8o45/28/), [4](https://jsfiddle.net/ps200306/zf3s8o45/33/), [5](https://jsfiddle.net/ps200306/zf3s8o45/31/)). But the extracted version only works in cases where either the keys change ([1](https://jsfiddle.net/ps200306/zf3s8o45/24/), [4](https://jsfiddle.net/ps200306/zf3s8o45/33/)) or only non-key fields on existing instances are updated ([5](https://jsfiddle.net/ps200306/zf3s8o45/31/)). In all five extracted cases I am _not_ passing the item into the child component as an initialisation parameter.

I retried all five cases, this time passing the item as an initialisation parameter to the extracted component, and they all seem to work. But this is kind of weird in itself. In case 1 and 2 where I replace either the array or the objects, the object that has changed is not one that I passed to the child component so I'd expect that not to work. (I should also mention that my real-world case is _not_ working when I pass the object to the child, though it's complicated and I need some more checking to make sure I'm not doing something wrong). Getting very confused here.

I'm thinking the easiest workaround (since I have something going live imminently) might be to just inject a sequentially generated artificial key into each of my objects. That means the keys are guaranteed to change every time. (EDIT: I can confirm that this works).

**Reply by ekwoka** [reply]
Inlining is meant to mean having the x-data be like `x-data="{ foo: 'bar' }"`


That also breaks it.

What does solve it is doing something like

`{ item }` (in the x data, where you pass the object in more explicitly.

**Reply by peterswords** [reply]
@ekwoka , then I'm misunderstanding. _All_ of my examples are inlined:

``` HTML
<body x-data="{
    my_array: [{x:'x', k:1},{x:'y', k:2}],
    click() {
        this.my_array = [{x:'a', k:3},{x:'b', k:4}];
    }
}">
```

The problem I'm running up against occurs when I try to use that data in a div that happens to be sandwiched inside another (extracted) component. The extracted component has no data, inline or otherwise.

Could you post a complete example of where you say inlining breaks it?

**Reply by ekwoka** [reply]
I mean for the looped component.

not the parent.

```html
 <template x-for="item in my_array" :key="item.k">
        <div x-data="{ foo: 'bar' }" >
            <div x-text="item.x"></div>
        </div>
    </template>
```

**Reply by peterswords** [reply]
Ahh, ok, now I get it. I thought the problem was that I was using a nested "extracted" component using `Alpine.data`. But now I see that it's caused by _any_ intervening component with its own `x-data`. And yes, I see that with this scenario changes to the outer array are only reflected if you change keys, not if you replace the entire array but don't change keys.

You are no doubt ahead of me here, but it would seem that the outer data context is somehow copied into the inner one and only certain types of change propagate from outer to inner. In any case, _something_ is different depending on whether there is an intervening context.

**Reply by ekwoka** [reply]
Yeah, it basically seems that the process of copying the outer looped scope to the inner does not happen on changes. It basically just does `is the key same? okay leave it` It might be a simple fix. I'll try to find it.

**Comment by miquelbrazil** [comment]
I feel like I'm experiencing an issue similar to this, but with `x-if`.

**Reply by ekwoka** [reply]
What is the issue? this shouldn't really affect if at all.

**Reply by miquelbrazil** [reply]
@ekwoka I had an `x-data` attribute on a parent element of two child components toggled by `x-if`. The child components contained input elements bound to a property on the parent `x-data`. They also had their own `x-data` attributes on the `<template>` tag surrounding them (the `<template>` tag only had a single child element as per the Alpine documentation). The initial value of the parent wasn't being reflected in the child component when it would get rendered according to the `x-if` rules. Also, when the child component's bound value was changed, this wasn't reflected in the parent.

I moved the `x-data` attribute in the child components to the inner element of the `<template>` tag instead of directly on it. This seemed to address the updated value when changed in the component but not the initial value that comes from the parent. I eventually assumed the issue was related to 1) the `<template>` tag not being a "real" part of the DOM causes the `x-data` attributes on it to produce a race condition when rendering or just unexpected mutations, and 2) I also figured the insertion and removal into the DOM had an impact on how those mutations bubbled up to the parent element. I ended up going with another strategy to solve the design but figured I'd share my shot-in-the-dark hypothesis anyway.

**Reply by ekwoka** [reply]
Yes, x-data shouldn't be on the template tag itself.

**Reply by shashank-brightness** [reply]
@ekwoka 
I am facing similar issue after initializing alpine manually in a shadow dom (without shadow dom works fine). I have also kept key in both templates outer and inner still when i try to change array, dom is not updating. If i only keep single loop and instead of `x-if` within 2nd loop, use as x-show on child div, it also works fine at that time as well.

```
<template x-for="shape in shapes" :key="shape.id">
  <template x-if="shape.type === 'motive'" :key="shape.id">
      <div class="col-4 mb-2  text-center position-relative editor-interface__hover__remove" :class="shape.hidden ? 'd-none': ''">
      </div>
  </template>
</template>
```

Also if I take one inner div before child template, it still works fine as well but this structure is not something i need. I need col-4 as direct child of upper template,

```
<template x-for="shape in shapes" :key="shape.id">
  <div>
    <template x-if="shape.type === 'motive'" :key="shape.id">
        <div class="col-4 mb-2  text-center position-relative editor-interface__hover__remove" :class="shape.hidden ? 'd-none': ''">
        </div>
    </template>
  </div>
</template>
```

but not with nested templates like the above shown in 1st example. Can you tell what is missing if first structure is within shadow dom and alpine was init manually as i said before?

**Reply by ekwoka** [reply]
You cannot put template as a child of a template.

So there are 2 ways to generally handle this

Recommended one for yours:

Just use a filter on for

`x-for="thing in things.filter(thing => thing.stuff = 'whatever')"`

Alternatively, a way to handle this and multi child templates in a way the layout can work with:

Add `display: contents` to the wrapping div of the inner template.

**Reply by shashank-brightness** [reply]
Okay I got it worked. Thanks for the quick help. I was not aware that we can even filter array within x-for itself in a syntax and also that root must not be template but anything else than that. May be better to add this example to [docs](https://alpinejs.dev/directives/for).

**Reply by ekwoka** [reply]
The rule about the templates is in the docs, I'm pretty sure, and Alpine expressions are JavaScript. So....you can write any JavaScript in them.

**Comment by ekwoka** [comment]
@peterswords 

Fix in #3504

**Reply by peterswords** [reply]
@ekwoka Many thanks
